pkg/git/git.go

```diff
@@ -0,0 +1,157 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package git
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/blang/semver/v4"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/storage/memory"
+)
+
+type Repo struct {
+	Ref           string
+	DefaultBranch string
+	Tags          []string
+	Branches      []string
+}
+
+func GetRepo(ref, url string) (*Repo, error) {
+	repo := new(Repo)
+	repo.Ref = ref
+
+	rem := git.NewRemote(memory.NewStorage(), &config.RemoteConfig{
+		Name: "origin",
+		URLs: []string{url},
+	})
+
+	refs, err := rem.List(&git.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ref := range refs {
+		if ref.Name().IsTag() {
+			repo.Tags = append(repo.Tags, ref.Name().Short())
+		} else if ref.Name().IsBranch() {
+			repo.Branches = append(repo.Branches, ref.Name().Short())
+		} else if ref.Name() == "HEAD" { // Default branch.
+			repo.DefaultBranch = ref.Target().Short()
+		}
+	}
+
+	return repo, nil
+}
+
+type RefType int
+
+const (
+	DefaultBranchRef RefType = iota
+	ReleaseBranchRef
+	ReleaseRef
+	NoRef
+)
+
+func (rt RefType) String() string {
+	if rt >= DefaultBranchRef && rt <= NoRef {
+		return [...]string{"Default Branch", "Release Branch", "Release", "No Ref"}[rt]
+	}
+	return ""
+}
+
+// BestRefFor Returns module@ref, isRelease
+func (r *Repo) BestRefFor(this semver.Version, ruleset RulesetType) (string, RefType) {
+
+	switch ruleset {
+	case AnyRule, ReleaseOrReleaseBranchRule, ReleaseRule:
+		var largest *semver.Version
+		// Look for a release.
+		for _, t := range r.Tags {
+			if sv, ok := normalizeTagVersion(t); ok {
+				v, _ := semver.Make(sv)
+				// Go does not understand how to fetch semver tags with pre or build tags, skip those.
+				if v.Pre != nil || v.Build != nil {
+					continue
+				}
+				if v.Major == this.Major && v.Minor == this.Minor {
+					if largest == nil || largest.LT(v) {
+						largest = &v
+					}
+				}
+			}
+		}
+		if largest != nil {
+			return fmt.Sprintf("%s@%s", r.Ref, tagVersion(*largest)), ReleaseRef
+		}
+	}
+
+	switch ruleset {
+	case AnyRule, ReleaseOrReleaseBranchRule, ReleaseBranchRule:
+		var largest *semver.Version
+		// Look for a release branch.
+		for _, b := range r.Branches {
+			if bv, ok := normalizeBranchVersion(b); ok {
+				v, _ := semver.Make(bv)
+
+				if v.Major == this.Major && v.Minor == this.Minor {
+					if largest == nil || largest.LT(v) {
+						largest = &v
+					}
+				}
+			}
+		}
+		if largest != nil {
+			return fmt.Sprintf("%s@%s", r.Ref, branchVersion(*largest)), ReleaseBranchRef
+		}
+	}
+
+	switch ruleset {
+	case AnyRule:
+		// Look for a Return default branch.
+		return fmt.Sprintf("%s@%s", r.Ref, r.DefaultBranch), DefaultBranchRef
+	}
+
+	// No ref found with the provided rule
+	return r.Ref, NoRef
+}
+
+func normalizeTagVersion(v string) (string, bool) {
+	if strings.HasPrefix(v, "v") {
```

Comment:
`if v[0] = 'v'` would probably work?

Reply:
you have to also check `if len(v) > 0 && v[0] == 'v' {` and it is easier to read as `HasPrefix`

Reply:
Do you actually expect non-empty strings at that point? It's a private function.

Reply:
it is processing things that the git call returned, the server could return an empty tag?

Reply:
I am gonna stick with readability over minor speed improvements, the QPS is low here.